Objects/dictobject.c

```diff
@@ -1737,7 +1736,44 @@ _PyDict_GetItem_KnownHash(PyObject *op, PyObject *key, Py_hash_t hash)
 
     ix = _Py_dict_lookup(mp, key, hash, &value);
     assert(ix >= 0 || value == NULL);
-    return value;
+    return value;  // borrowed reference
+}
+
+// Return the object from dictionary *op* which has a key *key*.
+// - If the key is present, set *pvalue to a new strong reference to the value
+//   and return 0.
+// - If the key is missing, set *pvalue to NULL and return 0 .
+// - On error, raise an exception and return -1.
+int
+PyDict_GetItemRef(PyObject *op, PyObject *key, PyObject **pvalue)
```

Comment:
Why not strongly typed, since it is a `PyDict_` API?

```suggestion
PyDict_GetItemRef(PyDictObject *op, PyObject *key, PyObject **pvalue)
```

Reply:
I would like to move the whole C API towards the only usage of ``PyObject*``. Currently, some functions expect a specific type like PyDictObject or PyTupleObject, whereas some others expect PyObject.

Moreover, existing functions use PyObject.

Reply:
I created python/devguide#1127 to try to find a consensus regarding guidelines for weakly/stronly typed APIs.

Reply:
AFAICS, this is the only unresolved issue; there are pros and cons for each variant (`PyObject *` vs. `PyDictObject *`). I'm not sure we can find consensus in the near future :(

Reply:
If the argument type is changed to ``PyDictObject*``, it means that all existing calls to PyDict_GetItem() and PyDict_GetItemWithError() will also have to cast their argument to ``PyDictObject*`` when the code is updated to use ``PyDict_GetItemRef()``.

Just one example: ``PyDict_New()`` returns a ``PyObject*``. It sounds wrong to me that the caller has more knowledge about the expected type than the official ``PyDict_New()`` API.

```c
PyObject *obj = PyDict_New():
PyDictObject *dict_for_getitemref = (PyDictObject*)obj;  // why??
(...)
PyObject *value;
if (PyDict_GetItemStringRef(dict_for_getitemref, "key", &value) < 0) ... errror ...
```

The whole C API of PyDict requires ``PyObject*`` type: PyDict_Size(), PyDict_Update(), PyDict_Values(), PyDict_SetDefault(), ...

**If** the C API would be rewritten from scratch, maybe it would make sense to use different types. But this PR is an **incremental fix** of the Python C API. Changing the parameter type would better fit into the idea of a brand new C API: https://github.com/capi-workgroup/problems/issues/55